Catch Exception in combine

TomAugspurger opened the pull request:
Closes https://github.com/pandas-dev/pandas/issues/22850

TomAugspurger committed:
Closes https://github.com/pandas-dev/pandas/issues/22850

pep8speaks commented:
Hello @TomAugspurger! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/core/series.py`](https://github.com/TomAugspurger/pandas/blob/fdd43c4de98992a6f97a835fdbb525f829ef1d69/pandas/core/series.py) !

 - There are no PEP8 issues in the file [`pandas/tests/extension/test_ops.py`](https://github.com/TomAugspurger/pandas/blob/fdd43c4de98992a6f97a835fdbb525f829ef1d69/pandas/tests/extension/test_ops.py) !

TomAugspurger committed:
Moves

TomAugspurger committed:
Imports

TomAugspurger committed:
Note

TomAugspurger commented:
The release note I just added is deliberately vague. I don't want non-EA authors poking around in here :)

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -505,6 +505,7 @@ ExtensionType Changes
 - :meth:`Series.astype` and :meth:`DataFrame.astype` now dispatch to :meth:`ExtensionArray.astype` (:issue:`21185:`).
 - Slicing a single row of a ``DataFrame`` with multiple ExtensionArrays of the same type now preserves the dtype, rather than coercing to object (:issue:`22784`)
 - Added :meth:`pandas.api.types.register_extension_dtype` to register an extension type with pandas (:issue:`22664`)
+- Renamed test prototype module names to avoid conflict with modules from the standard library (:issue:`22936`)
```
not clear what is changing here

TomAugspurger replied:
https://github.com/pandas-dev/pandas/pull/22936#issuecomment-426240558

jreback commented on pandas/tests/extension/test_ops.py:
```diff
@@ -0,0 +1,20 @@
+from pandas.tests.extension.decimal_array import DecimalArray
+
+
+def test_combine_from_sequence_raises():
```
why is this not with the existomg tests?

TomAugspurger replied:
This isn't a base extensionarray test that applies to all EAs. It's just to EAs that may raise errors other than TypeError in `_values_from_sequence`.

We need a place for simple tests like this that

1. Aren't applicable to every downstream ExtensionArray (so not in `tests/extension/base`)
2. Are testing pandas dispatching, rather than the behavior of a specific EA (so not in `tests/arrays/`)

This seemed like a sensible home.

jorisvandenbossche commented on pandas/tests/extension/decimal_array/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from .array import DecimalArray, DecimalDtype
+
+
+__all__ = ['DecimalArray', 'DecimalDtype']
```
can you add here make_data ? (that are typically the ones that I use to test things out locally)

jbrockmendel commented:
+1 on the renaming; need caffeine for the rest

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into combine-exception

TomAugspurger committed:
api

TomAugspurger committed:
move back

TomAugspurger committed:
move test

TomAugspurger committed:
handle test

TomAugspurger committed:
remove old note

TomAugspurger committed:
fixed move

TomAugspurger committed:
fixed move

TomAugspurger commented:
Updates:

1. Moved the test with the rest extension ops tests in `test_decimal.py` (though I still think we need a home for non-EA specific, non-interface tests). Maybe that's just `pandas/tests/test_ops.py` though. 
2. Undid the module renames (not a bad idea to do on its own though)
3. Changed ExtensionScalarOpsMixin to catch exceptions in `_from_sequence`, and return an ndarray in that case. Added a test for that (previously untested AFAICT). Docs coming in #22932 

This and https://github.com/pandas-dev/pandas/pull/22932 (divmod) are both touching the same lines now, sorry. I'd recommend merging this one first.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into combine-exception

jbrockmendel commented on pandas/core/series.py:
```diff
@@ -2324,9 +2324,11 @@ def combine(self, other, func, fill_value=None):
             try:
                 new_values = self._values._from_sequence(new_values)
-            except TypeError:
+            except Exception:
```
Can you add a comment here linking to this PR.  Inevitably I'm going to forget why this isn't catching something more specific.

jorisvandenbossche commented on pandas/tests/extension/decimal/test_decimal.py:
```diff
@@ -275,3 +271,34 @@ def test_compare_array(self, data, all_compare_operators):
+
+
+def test_scalar_ops_from_sequence_raises():
+    arr = DecimalArrayWithoutFromSequence([
```
can you add a comment here to explain that if from_sequence raises, the result should be a normal numpy array ?

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into combine-exception

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -777,8 +777,8 @@ def convert_values(param):
-                except TypeError:
-                    pass
+                except Exception:
+                    res = np.asarray(res, dtype=object)
```
Hmm, this looks like it's in the wrong place.

Right now with `coerce_to_dtype=False` we'd return an list, not an ndarray. I assume we want an ndarray of objects regardless of `coerce_to_dtype`? I don't want the return type to be `Union[ExtensionArray, List, ndarray]`

TomAugspurger replied:
Grr... ok what do to here?

If we do

```
                res = np.asarray(res, dtype=object)
```

we'll start failing some other tests like `Series[DecimalArray] == 1`, since we get an object dtype instead of bool. So we just leave it up to NumPy's regular inference then, `res = np.asarray(res)`?

TomAugspurger committed:
Always return ndarray

TomAugspurger commented:
Updates:

1. Return an ndarray instead of a list, even when `coerce_to_dtype=False`
2. Documented the expected return type. Are people familiar with the mypy syntax? `-> Callable[[Any, Any], Union[ndarray, ExtenionsionArray]` means the return value is a callable taking two arguments. Right now that's just Any and Any, since there could be things like `EA + 1`. All we know is that one of them is an EA. The last item in `Callable` is the return value of the callable, which is either an ndarray or EA.
3. Added links back here around the broad `except Exception`s

jreback commented:
thanks @TomAugspurger 

not sure I completely understand the use of ``UserDict`` in the json/array, but you did leave a TODO

TomAugspurger commented on pandas/tests/extension/json/array.py:
```diff
@@ -179,3 +181,10 @@ def _values_for_argsort(self):
+
+
+def make_data():
+    # TODO: Use a regular dict. See _NDFrameIndexer._setitem_with_indexer
```
@jorisvandenbossche requested these be exported, since they're useful for creating datasets. It'd be more strange to import it from the test file.

TomAugspurger commented:
The UserDict goes back to the original implementation. We have a few `if isinstance(thing, dict)` internally that break things for `JSONArray`. We'll need to sort that if / when we properly support nested data, where the "scalar" type for an array is actually a sequence like a dict or array.
